pkg/client/client_v2.go

```diff
@@ -581,6 +602,80 @@ func (c *clientV2) tryPerformAction(actionResults chan *proto.ActionResponse, ac
 	}()
 }
 
+func (c *clientV2) tryPerformDiagnostics(actionResults chan *proto.ActionResponse, action *proto.ActionRequest) {
+	// find the unit
+	c.unitsMu.RLock()
+	unit := c.findUnit(action.UnitId, UnitType(action.UnitType))
+	c.unitsMu.RUnlock()
+	if unit == nil {
+		actionResults <- &proto.ActionResponse{
+			Token:  c.token,
+			Id:     action.Id,
+			Status: proto.ActionResponse_FAILED,
+			Result: ActionErrUnitNotFound,
+		}
+		return
+	}
+
+	// gather diagnostics hooks for this unit
+	diagHooks := make(map[string]diagHook)
+	c.dmx.RLock()
+	for n, d := range c.diagHooks {
+		diagHooks[n] = d
+	}
+	c.dmx.RUnlock()
+
+	// unit hooks after client hooks; allows unit specific to override client hooks
+	unit.dmx.RLock()
+	for n, d := range unit.diagHooks {
+		diagHooks[n] = d
+	}
+	unit.dmx.RUnlock()
+
+	// perform diagnostics in goroutine, so we don't block other work
+	go func() {
+		res := make([]*proto.ActionDiagnosticUnitResult, 0, len(diagHooks))
+		for n, d := range diagHooks {
+			content := d.hook()
```

Comment:
Is it worth considered hooks that could block or time out? Or do we expect them to always complete quickly? Is it worth passing a done channel or a context through to the hooks?

Reply:
I think we should leave that up to the hook. I think encoding the timeout at this level would mean you have to register with a timeout, but if a hook needs to manage that timeout it could just do that in its own function.

I would hope that most are really fast and can quickly get the results back to the Elastic Agent. The Elastic Agent will place a timeout on the Diagnostic request to the component so if it takes longer than that timeout to return then the Elastic Agent will just continue on.

Reply:
Thanks I just misread this, I was focused on `dmx.RLock()` and missed that it was `unit.dmx.RLock()` on the second loop